### DISCUSSION TITLE
Sorting Challenge with MongoDB CRUD: Seeking Guidance

### DISCUSSION THREAD

**Discussion by Noste92** [mia-platform/community#349] [Q&A] [answered]
### Which section?

Marketplace

### Which is the version?

CRUD 6.3.0

### Description

Hello everyone,

We're currently grappling with an unexpected behavior in our CRUD component, specifically when sorting a MongoDB collection. The issue arises when sorting on a property that shares the same value across multiple documents, causing anomalies during pagination, with some items missing and others repeating.

Our research points to a potential problem with MongoDB's sort, skip, and limit functions, particularly when dealing with identical property values across documents. One common suggestion is to introduce a secondary sort on a unique property, such as _id. Here's the references:
- [Mongo Community discussion](https://www.mongodb.com/community/forums/t/sorting-data-causes-duplication/234217/3)
- [An old stackoverflow discussion](https://stackoverflow.com/questions/44692075/mongo-db-duplication-issue-while-using-sorting-with-limit-and-skip-in-aggregatio)

Our struggle is centered between the backkit client and the CRUD. If anyone has encountered a similar challenge or has insights on a possible solution, we'd greatly appreciate your input.

### Actual Outcome

Sorting anomaly affecting data retrieval between the client backkit and CRUD due to identical property values.

### Expected Outcome

Seeking community advice and potential solutions to address the MongoDB sorting issue.

**Comment by frengooo** [comment]
@flower-of-the-bridges should have some information about this topic.
We had a similar issue on a client, there is a discussion too:
https://github.com/mia-platform/community/discussions/346

**Reply by flower-of-the-bridges** [reply]
Hi @silversoul93 @Noste92 

Yes, I've encountered some issues with pagination on MongoDB collections. My advice is to use the sort parameter `_s` of the CRUD service on a primary key (for example, the `_id` field could be enough),  to avoid record duplications on different calls. 

I think you can't provide with the back-kit crud client such property, but you can specify in [a table](https://docs.mia-platform.eu/docs/microfrontend-composer/back-kit/components/table#properties--attributes) the property `initialSortProperty` to inject the sort field into the CRUD Service request

**Reply by Noste92** [reply]
Hi @flower-of-the-bridges,

We actually gave initialSortProperty a shot, and it does work for the initial page load. However, when users want to sort by a different field, we're back to facing the same duplication issue.

Ideally, we'd love to implement a solution where sorting always happens on two fields – the selected field and a secondary one like `_id`. This would prevent duplicates and keep things smooth. The big question for us is where to insert this behavior. 
Would it make sense as a feature in back-kit or should it be tackled at the CRUD level?

**Reply by flower-of-the-bridges** [reply]
Hi @Noste92 

The thing is, like you said, the sort should be done on two fields: `_id` and the field selected by the user.

The CRUD service can already have [sort on multiple fields](https://docs.mia-platform.eu/docs/runtime_suite/crud-service/overview_and_usage#sort), so I think back-kit, given the behavior of pagination, should append the sort property chosen by the user to the `_id` one.

**Reply by clabene** [reply, marked as answer]
As mentioned, atm the `bk-crud-client` does not support sorting by multiple fields, nor does it support specifying default values for query parameters (like `_s`).

Feel free to create a new feature request to extend `bk-crud-client` functionalities to address the discussed scenario.

Thanks!

**Reply by Noste92** [reply]
That's clear to me. Thank you guys for your support!
